### DISCUSSION TITLE
umi4 qiankun : No routes matched location "/app1/system"

### DISCUSSION THREAD

**Discussion by xzz-1005** [umijs/umi#12362] [Q&A]
前提：qiankun umi4主应用+umi4子应用
 1. 主应用设置路由为
{
      name: '子应用1',
      path: '/app1',
      microApp: 'app1',
      icon: 'TabletOutlined',
      routes: [
          {
            name: '系统管理',
            path: '/app1/system',
          },
      ]
    },
 报错No routes matched location "/app1/system" 
 2. 主应用设置路由为以下，可以跳转app1正常展示/app1/system页面
  {
      name: 'app1',
      path: '/app1/*',
      microApp: 'app1',
    },
    问题：为啥方法1设置报错

**Comment by fz6m** [comment]
一般情况你是不知道子应用有哪些路径的，因为子应用的路径可能未来新增，这个时候父应用如果还要再改一遍路由表发布的话，解耦的意义就不大了，所以用通配 `path: '/app1/*',` 是比较好的，之后如果你用了 max 自带的 layout 的话，可以在 layout 的 menu.request 里自己改菜单，代码参考见 https://github.com/umijs/umi/issues/12257#issuecomment-2039823344 ，文档见 [ProLayout](https://procomponents.ant.design/components/layout#menu) 。

如果你打算细分路由表的话，还是用 [`<MicroApp  />`](https://umijs.org/docs/max/micro-frontend#microapp--%E7%BB%84%E4%BB%B6%E5%BC%95%E5%85%A5%E5%AD%90%E5%BA%94%E7%94%A8) 手动渲染子应用吧，把这些路由的 component 配成这个组件就行了。

至于在静态路由（`.umirc.ts`）里配路由表还是在 [`patchClientRoutes`](https://umijs.org/docs/api/runtime-config#patchclientroutes) 里动态改路由表，都可以。

这里还有一个复杂的例子可以参考：https://github.com/umijs/umi/issues/12277#issuecomment-2065597766

**Reply by xzz-1005** [reply]
这里还有一个复杂的例子可以参考：https://github.com/umijs/umi/issues/12277#issuecomment-2065597766
仔细看了一下这个例子，我好像遇到跟他同样问题，其实就是想在主应用里直接展示子应用的菜单，而不是点进去单独展示子应用菜单，我这样写就解决了
{
name: '子应用1',
path: '/app1/*',
microApp: 'app1',
icon: 'TabletOutlined',
routes: [
{
name: '系统管理',
path: 'system',
},
]
},

**Reply by wyx2014** [reply]
大佬，我想通过主应用控制子应用路由，虽然menu上可以隐藏，但依然可以通过路由打开没有权限的路由页面。能给点思路吗？

**Comment by xzz-1005** [comment]
感谢！大佬！~~~~~~
